tmpl/wiki/package-hardening.md

```diff
@@ -0,0 +1,118 @@
+`mirage lint`) but to start we list the criteria that such an hardened
+package should follow.
+
+### Project Paperwork
+
+- Have well-identified maintainers. GitHub recently
+  [introduced code owners](https://github.com/blog/2392-introducing-code-owners),
+  which may be worth to use for some (of our core) libraries.
+
+- Have a design document specifying the scope of the library (and esp. what is
+  out of scope)
+
+- Use the correct set of tags in opam metadata
+
+- Use proper names: same name for opam and ocamlfind package, and also ideally
+  for the top-level module (see next point).
+
+- Use module alias instead of packs to define proper namespaces
+  (to avoid name-clashes)
+
+- Have documentation
+
+- Have a clear LICENSE file
+
+- Use [topkg](https://github.com/dbuenzli/topkg) to release, in order to ease
+  the task of the release manager(s).
+
+- Have proper indentation (using ocp-indent + checked-in ocp-indent file)
+
+- Use at most 80 columns (`ocp-indent` unfortunately doesn't check this)
+
+- Follow [OCaml programming guidelines](https://ocaml.org/learn/tutorials/guidelines.html)
+
+- Ideally, use [Jbuilder](https://github.com/janestreet/jbuilder) to build,
+  in order to have short feedback loops when fixing bugs spanning multiple
+  packages.
+
+### Code Good Practice
+
+- Provide pretty printers by using [fmt](http://erratique.ch/software/fmt/doc/Fmt.html)
+
+- Do not expose any exceptions in the public interface, but use the
+  [result](http://caml.inria.fr/pub/docs/manual-ocaml/libref/Pervasives.html#TYPEresult)
+  type and [rresult](http://erratique.ch/software/rresult/doc/Rresult.html)
+  combinators!
+
+- Use [logs](http://erratique.ch/software/logs/doc/Logs.html) for logging
+
+- Use [astring](http://erratique.ch/software/astring/doc/Astring.html) for
+  string processing
+
+- Use [bos](http://erratique.ch/software/bos/doc/Bos.html) for operating system
+  interaction
+
+- Use [fpath](http://erratique.ch/software/fpath/doc/Fpath.html) for file paths
+
+- Avoid integer overflows (basically every addition and subtraction, as well
+  as multiplication needs to be guarded unless you know that an overflow can
+  never happen (in this case, a comment should be suggested))
+
+- Work on 32bit (esp. in regards to the last point)
+
+- Avoid global mutable state
+
+- Be aware of non-tail recursive functions
+
+- Be aware of polymorphic equality drawbacks -- when possible define your own
+  specialized functions.
+
+### Project Confidence
+
+- All direct dependencies are specified in the opam file (e.g. sometimes some
+  direct dependencies are included by transitivity -- this is not very stable
+  and could break the package in the future if one of a transitive dependency
+  is updated)
+
+- Use `-safe-string` (looks like this will be the default in 4.06)
+
+- Have unit tests (using alcotest) (ideally with coverage report)
+
+- Have fuzz tests (using crowbar)
```

Comment:
I think this is a great idea for all projects, but the tooling really isn't ready for this to be included yet in a document people might be trying to adhere to.  "Have randomized property-based testing" is more achievable and will give people the right foundation to use tools like Crowbar later, once they're more ready to be integrated in a CI pipeline.